gcloud/src/main/java/org/jimmutable/gcloud/search/IndexManager.java

```diff
@@ -0,0 +1,167 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.jimmutable.gcloud.logging.LogSupplier;
+
+import com.google.appengine.api.search.Document;
+import com.google.appengine.api.search.Field;
+import com.google.appengine.api.search.Index;
+import com.google.appengine.api.search.IndexSpec;
+import com.google.appengine.api.search.OperationResult;
+import com.google.appengine.api.search.PutException;
+import com.google.appengine.api.search.PutResponse;
+import com.google.appengine.api.search.Results;
+import com.google.appengine.api.search.ScoredDocument;
+import com.google.appengine.api.search.SearchException;
+import com.google.appengine.api.search.SearchServiceFactory;
+import com.google.appengine.api.search.StatusCode;
+
+/**
+ * Use this class to upsert, delete and get a document from a search Index
+ * 
+ * @author trevorbox
+ *
+ */
+public class IndexManager {
+
+	private static final Logger logger = Logger.getLogger(IndexManager.class.getName());
+
+	private static Index getIndex(IndexId indexId) {
+		IndexSpec indexSpec = IndexSpec.newBuilder().setName(indexId.getSimpleValue()).build();
+		return SearchServiceFactory.getSearchService().getIndex(indexSpec);
+	}
+
+	/**
+	 * Performs a get operation on the index for the given DocumentId
+	 * 
+	 * @param index_id
+	 *            IndexId
+	 * @param document_id
+	 *            DocumentId
+	 * @param default_value
+	 *            The Document to return in the event the document is not found
+	 * @return
+	 */
+	public static Document getComplex(IndexId index_id, DocumentId document_id, Document default_value) {
+		Index index = getIndex(index_id);
+		Document d = index.get(document_id.getSimpleValue());
+		if (d != null) {
+			return d;
+		}
+		return default_value;
+	}
+
+	/**
+	 * Performs a deletion on a DocumentId within an IndexId and will retry up to
+	 * three times in the even a runtimeException is thrown.
+	 * 
+	 * @param index_id
+	 *            IndexId
+	 * @param document_id
+	 *            DocumentId
+	 * @throws InterruptedException
+	 */
+	public static void delete(IndexId index_id, DocumentId document_id) throws InterruptedException {
+		Index index = getIndex(index_id);
+
+		final int maxRetry = 3;
+		int attempts = 0;
+		int delay = 2;
+		while (attempts < maxRetry) {
+			attempts++;
+			try {
+				index.delete(index_id.getSimpleValue());
+				logger.info(new LogSupplier("Deleted documentId %s in indexId %s", index_id.getSimpleValue(),
+						document_id.getSimpleValue()));
+				break;
+			} catch (RuntimeException e) {
+				logger.log(Level.SEVERE, e, new LogSupplier("Failed to delete documentId %s in indexId %s"));
+				Thread.sleep(delay * 1000);
+				delay *= 2; // easy exponential backoff
+			}
+		}
+	}
+
```

Comment:
signature should (a) throw no exceptions but (b) return a boolean (success or failure).  Far easier to work with in practice and consistent with other parts of our API

Need validation on params: Validator.notNull(index_id, document_id)